projects/fluid_simulation/sim.h

```diff
@@ -62,208 +85,128 @@ struct Bouncer
-    // Parent application data.
-    ppx::Application* mApp;
-
-    // Device to use.
-    ppx::grfx::DevicePtr mDevice;
-
-    // Resolution to use for the simulation.
-    ppx::uint2 mResolution;
-
     // Simulation parameters.
     SimulationConfig mConfig;
 
     // Frame synchronization data.
     std::vector<PerFrame> mPerFrame;
 
     // Descriptor pool for compute and graphics shaders.
-    ppx::grfx::DescriptorPoolPtr mDescriptorPool;
-
-    // Compute resources (pipeline interface, descriptor layout, sampler, etc).
-    ComputeResources mCompute;
-
-    // Graphics resources (pipeline interface, descriptor layout, sampler, etc).
-    GraphicsResources mGraphics;
-
-    // Textures used for filtering.
-    std::unique_ptr<Texture>              mBloomTexture;
-    std::vector<std::unique_ptr<Texture>> mBloomTextures;
-    std::unique_ptr<Texture>              mCheckerboardTexture;
-    std::unique_ptr<Texture>              mCurlTexture;
-    std::unique_ptr<Texture>              mDisplayTexture;
-    std::unique_ptr<Texture>              mDitheringTexture;
-    std::unique_ptr<Texture>              mDivergenceTexture;
-    std::unique_ptr<Texture>              mDrawColorTexture;
-    std::unique_ptr<Texture>              mDyeTexture[2];
-    std::unique_ptr<Texture>              mPressureTexture[2];
-    std::unique_ptr<Texture>              mSunraysTempTexture;
-    std::unique_ptr<Texture>              mSunraysTexture;
-    std::unique_ptr<Texture>              mVelocityTexture[2];
+    ppx::grfx::DescriptorPoolPtr mDescriptorPool = nullptr;
+
+    // Pipeline interface, descriptor layout and sampler used by compute shaders.
+    ppx::grfx::PipelineInterfacePtr   mComputePipelineInterface   = nullptr;
+    ppx::grfx::DescriptorSetLayoutPtr mComputeDescriptorSetLayout = nullptr;
+    ppx::grfx::SamplerPtr             mClampSampler               = nullptr;
+    ppx::grfx::SamplerPtr             mRepeatSampler              = nullptr;
+
+    // Pipeline interface, descriptor layout and other resources used for rendering textures.
+    ppx::grfx::PipelineInterfacePtr   mGraphicsPipelineInterface   = nullptr;
+    ppx::grfx::DescriptorSetLayoutPtr mGraphicsDescriptorSetLayout = nullptr;
+    ppx::grfx::GraphicsPipelinePtr    mGraphicsPipeline            = nullptr;
+    ppx::grfx::VertexBinding          mGraphicsVertexBinding;
+
+    // Grids used for filtering.
+    std::vector<std::unique_ptr<SimulationGrid>> mBloomGrids;
+    std::unique_ptr<SimulationGrid>              mBloomGrid      = nullptr;
+    std::unique_ptr<SimulationGrid>              mCurlGrid       = nullptr;
+    std::unique_ptr<SimulationGrid>              mDisplayGrid    = nullptr;
+    std::unique_ptr<SimulationGrid>              mDitheringGrid  = nullptr;
+    std::unique_ptr<SimulationGrid>              mDivergenceGrid = nullptr;
+    std::unique_ptr<SimulationGrid>              mDrawColorGrid  = nullptr;
+    std::unique_ptr<SimulationGrid>              mDyeGrid[2];
```

Comment:
Might want to replace the 2 with a named constant of your choice in case someone else is reading this in the future knows what it means.

Reply:
Done.